File: oak_restricted_kernel_interface/src/syscalls.rs
//
// Copyright 2022 The Project Oak Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

use bitflags::bitflags;
use strum::FromRepr;

/// Syscalls above this are unsafe, their behavior and availability cannot be
/// relied upon.
pub const UNSTABLE_SYSCALL_SPACE: usize = i32::MAX as usize - (10 ^ 3);

/// System calls implemented by Oak Restricted Kernel.
///
/// In general, the system calls are inpired by and look similar to the
/// Linux/POSIX ABI, but we make no claims about adhering to the behaviour, or
/// specification, of either of those.
///
/// See <https://github.com/torvalds/linux/blob/master/arch/x86/entry/syscalls/syscall_64.tbl> for a
/// full list of system call numbers; we will ever only support a small subset
/// of these.
#[repr(usize)]
#[derive(Debug, FromRepr)]
pub enum Syscall {
    /// Read from a file descriptor.
    ///
    /// Arguments:
    ///   - arg0 (c_ssize_t): file descriptor number
    ///   - arg1 (*mut c_void): pointer to the buffer to be filled
    ///   - arg2 (c_size_t): size of the buffer
    /// Returns:
    ///   a value of <errno::Errno> on failure; otherwise, the number of bytes
    /// read.
    Read = 0,

    /// Write to a file descriptor.
    ///
    /// Arguments:
    ///   - arg0 (c_ssize_t): file descriptor number
    ///   - arg1 (*const c_void): pointer to the buffer containing data to be
    ///     written
    ///   - arg2 (c_size_t): size of the buffer
    /// Returns:
    ///   a value of <errno::Errno> on failure; otherwise, the number of bytes
    /// written.
    Write = 1,

    /// Creates a mapping for memory.
    /// Arguments:
    ///   - arg0 (*const c_void): hint for start address for the new mapping,
    ///     may be nullptr
    ///   - arg1 (c_size_t): size of the new mapping
    ///   - arg2 (c_int): protection on mapping (PROT_EXEC, PROT_READ,
    ///     PROT_WRITE, PROT_NONE)
    ///   - arg3 (c_int): flags. We require MAP_PRIVATE and MAP_ANONYMOUS to be
    ///     set, and additionally support MAP_FIXED.
    ///   - arg4 (c_int): file descriptor. Ignored, as we only support anonymous
    ///     mappings. Should be set to -1 by caller.
    ///   - arg5 (c_int): offset. Ignored, as we only support anonymous
    ///     mappings. Should be set to 0 by caller.
    /// Oak Restricted Kernel considerations:
    ///   - our mmap will work on 2 MiB chunks; even if size is 1 byte, we will
    ///     reserve a 2 MiB chunk of memory. Thus, size should be kept as a
    ///     multiple of 2 MiB.
    ///   - related to previous, the allocation address will always be 2
    ///     MiB-aligned (rounded upward from hint).
    ///   - MAP_FIXED requires address to be 2 MiB-aligned, and will return an
    ///     error if it'd touch any existing mappings.
    ///   - We do not support PROT_NONE; PROT_READ is always implied.
    Mmap = 9,

    /// Terminates he calling process.
    /// Arguments:
    ///   - arg0 (c_int): error code
    /// Oak Restricted Kernel considerations:
    ///   We don't expect the user process to terminate, so this triggers a
    /// kernel panic, no matter   the error code.
    Exit = 60,

    /// Flush a file descriptor.
    /// Arguments:
    ///   - arg0 (c_ssize_t): file descriptor number. Ignored by the kernel as
    ///     we don't support multiple file descriptors.
    /// Returns:
    ///   a value of <errno::Errno> on failure; 0, otherwise.
    Fsync = 74,

    /// Create new process from ELF file, without starting it.
    ///
    /// Arguments:
    ///   - arg0 (*mut c_void): pointer to the a buffer holding an ELF file
    ///   - arg1 (c_size_t): size of the buffer
    /// Returns:
    ///   a value of <errno::Errno> on failure; Otherwise the PID of the new
    /// process.
    UnstableCreateProcess = UNSTABLE_SYSCALL_SPACE,

    /// Switch the active execution execution to the process with the provided
    /// PID.
    ///
    /// Arguments:
    ///   - arg0 (c_size_t): PID of the process.
    /// Returns:
    ///   a value of <errno::Errno> on failure; 0, otherwise.
    UnstableSwitchProcess = UNSTABLE_SYSCALL_SPACE + 1,
}

bitflags! {
    #[repr(C)]
    pub struct MmapProtection: i32 {
        /// Pages may be read.
        const PROT_READ = 0x1;

        // Pages may be written.
        const PROT_WRITE = 0x2;

        /// Pages may be executed.
        const PROT_EXEC = 0x4;
    }
}

bitflags! {
    #[derive(Debug)]
    #[repr(C)]
    pub struct MmapFlags: i32 {
        /// Private copy-on-write mapping.
        const MAP_PRIVATE = 0x02;

        /// Don't interpret addr as a hint, but require mapping at given address.
        const MAP_FIXED = 0x10;

        /// The mapping is not backed by any file; contents are initialized to zero.
        const MAP_ANONYMOUS = 0x20;
    }
}

File: toolchain/newlib/syscalls.c
//
// Copyright 2023 The Project Oak Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

#include <errno.h>
#include <reent.h>
#include <sys/stat.h>
#include <sys/types.h>

static long __syscall0(long n) {
  unsigned long ret;
  asm volatile("syscall" : "=a"(ret) : "a"(n) : "rcx", "r11", "memory");
  return ret;
}

static long __syscall3(long n, long a1, long a2, long a3) {
  unsigned long ret;
  asm volatile("syscall" : "=a"(ret) : "a"(n), "D"(a1), "S"(a2), "d"(a3) : "rcx", "r11", "memory");
  return ret;
}

static long __syscall6(long n, long a1, long a2, long a3, long a4, long a5, long a6) {
  unsigned long ret;
  register long r10 asm("r10") = a4;
  register long r8 asm("r8") = a5;
  register long r9 asm("r9") = a6;
  asm volatile("syscall"
               : "=a"(ret)
               : "a"(n), "D"(a1), "S"(a2), "d"(a3), "r"(r10), "r"(r8), "r"(r9)
               : "rcx", "r11", "memory");
  return ret;
}

void _exit(int rc) {
  __syscall0(60);
  for (;;) {
  }
}

ssize_t _read_r(struct _reent* reent, int fd, void* buf, size_t len) {
  int ret = __syscall3(0, fd, (long)buf, len);

  if (ret < 0) {
    reent->_errno = ret;
    return -1;
  }
  return ret;
}

ssize_t _write_r(struct _reent* reent, int fd, const void* buf, size_t len) {
  int ret = __syscall3(1, fd, (long)buf, len);

  if (ret < 0) {
    reent->_errno = ret;
    return -1;
  }
  return ret;
}

ssize_t _lseek_r(struct _reent* reent, int fd, _off_t ptr, int dir) {
  reent->_errno = ENOSYS;
  return -1;
}

void* mmap(void* addr, int len, int prot, int flags, int fd, int offset) {
  return __syscall6(9, (long)addr, len, prot, flags, fd, offset);
}

void* _sbrk_r(struct _reent* reent, ptrdiff_t incr) {
  extern char end asm("_end");  // symbol provided by the linker
  static char* heap_end;

  if (heap_end == NULL) {
    // If we haven't initialized the heap at all yet, let's start at the next 2 MiB page boundary
    // after `_end`.
    heap_end = &end;

    if ((int)heap_end % 0x200000 != 0) {
      heap_end += 0x200000 - (int)heap_end % 0x200000;
    }
  }

  // We only increment in chunks of 2 MiB.
  if (incr % 0x200000 != 0) {
    incr += 0x200000 - incr % 0x200000;
  }

  char* prev_heap_end = heap_end;

  if (mmap(heap_end, incr,
           0x1 | 0x2,           // PROT_READ | PROT_WRITE
           0x02 | 0x10 | 0x20,  // MAP_PRIVATE | MAP_FIXED | MAP_ANONYMOUS
           -1, 0) != heap_end) {
    reent->_errno = ENOMEM;
    return -1;
  }

  heap_end += incr;

  return prev_heap_end;
}

int _open_r(struct _reent* reent, const char* file, int flags, int mode) {
  reent->_errno = ENOSYS;
  return -1;
}

int _close_r(struct _reent* reent, int fd) {
  reent->_errno = ENOSYS;
  return -1;
}

int _fstat_r(struct _reent* reent, int fd, struct stat* s) {
  reent->_errno = ENOSYS;
  return -1;
}

int _isatty_r(struct _reent* reent, int fd) {
  reent->_errno = ENOSYS;
  return -1;
}

int _getpid_r(struct _reent* reent) {
  reent->_errno = ENOSYS;
  return -1;
}

int _kill_r(struct _reent* reent, int pid, int signal) {
  reent->_errno = ENOSYS;
  return -1;
}

int getentropy(void* ptr, size_t n) { return -1; }
